### PR TITLE
DetectChunk Worker

### PR DESCRIPTION
**Pull request by zricethezav**
Bringing some optimizations @ahrav made into OSS.

Still want to run a few tests. Putting this as a draft for now

### Rails (Old, --no-verification)
```
real 27.28s
user 54.64s
sys  5.81s
memory:717728KB
cpu 221%
```

### Rails (New, --no-verification)
```
real 33.55s
user 81.87s
sys  7.17s
memory:770860KB
cpu 265%
```

### Rails (New, --only-verified)
```
real 33.84s
user 80.82s
sys  7.14s
memory:748548KB
cpu 259%
```

### Rails (Old, --only-verified)
```
real 26.73s
user 52.89s
sys  5.48s
memory:757616KB
cpu 218%
```

### Gitlab (Old, --only-verified)
```
real 1134.83s
user 681.88s
sys  104.51s
memory:3480160KB
cpu 69%
```

### Gitlab (New, --only-verified)
```
real 1231.04s
user 1084.24s
sys  121.95s
memory:3350048KB
cpu 97%
```

### CODE REVIEW

File: pkg/engine/engine.go [hunk 1]
```diff
@@ -242,7 +266,55 @@ func (e *Engine) DetectorAvgTime() map[string][]time.Duration {
...
+			var avgTime []time.Duration
+			if ok {
+				avgTime, ok = avgTimeI.([]time.Duration)
+				if !ok {
+					continue
+				}
+			}
+			e.detectorAvgTime.Store(detectorName, avgTime)
+		}
+		data.wgDoneFn()
```
Comment (mcastorina) [inline pkg/engine/engine.go:313]: The `continue` codepath means `data.wgDoneFn()` doesn't get called. I think the `if` statements could be refactored to:

```go
avgTimeI, ok := e.detectorAvgTime.Load(detectorName)
if ok {
    if avgTime, ok := avgTimeI.([]time.Duration); ok {
        e.detectorAvgTime.Store(detectorName, avgTime)
    }
}
```

**Review by mcastorina — COMMENTED** [review]
lgtm, just had one concern

**Comment by mcastorina** [comment]
Also, glancing at the benchmarks you posted, it looks like the `real` time went up? Is that expected?

**Comment by zricethezav** [comment]
@mcastorina no it's not. closing this for now. attempted to optimize this with a few different approaches including using errgroup. time still went up

**Pull request closed without merging**